### PR TITLE
Add Test Definition for OTPValidationAPI

### LINKED ISSUE #57: Provide test definition artefacts
**Problem description**
Provide test definition for this API

**Expected action**
Provide a Gherkin file with Test Definition

**Additional context**
To be done accordingly to https://github.com/camaraproject/Commonalities/pull/203

### PR DESCRIPTION
**Pull request by bigludo7**
Add Test Definition for OTPValidationAPI

#### What type of PR is this?

Add one of the following kinds:

* enhancement/feature

#### What this PR does / why we need it:

Add Test Definition for OTPValidationAPI

#### Which issue(s) this PR fixes:

Fixes #57 

#### Special notes for reviewers:

Need agreement on https://github.com/camaraproject/Commonalities/pull/203 before to be fully validated
Adding @jlurien as reviewer to check alignement with the guideline.

#### Changelog input

```
 release-note

```

#### Additional documentation 

This section can be blank.

```
docs

```

### CODE REVIEW

**Review by fernandopradocabrillo — COMMENTED** [review]
@bigludo7 
I think the scenarios that are included are fine. I kind of like it more separated in 2 different feature files one for each operation but I can live with it being in only one 😄 

I leave you a list with a few scenarios that I think are missing and if we target for an stable version we should include them.

/send-code:
1. Reached max number of codes that can be requested for a phone number -> ONE_TIME_PASSWORD_SMS.MAX_OTP_CODES_EXCEEDED
2. The phone number specified is not allowed to receive messages -> ONE_TIME_PASSWORD_SMS.PHONE_NUMBER_NOT_ALLOWED
3. The phone number specified does not belong to the operator -> ONE_TIME_PASSWORD_SMS.PHONE_NUMBER_NOT_ALLOWED
4. The phone number specified is a landline -> ONE_TIME_PASSWORD_SMS.PHONE_NUMBER_NOT_ALLOWED
5. The phone number specified is blocked for any business reason -> ONE_TIME_PASSWORD_SMS.PHONE_NUMBER_BLOCKED
6. The property message is missing -> INVALID_ARGUMENT
7. The property phone number is missing -> INVALID_ARGUMENT
8. The message specified reached a max length stablished by the operator -> INVALID_ARGUMENT

/validate-code:
1. Verification expired because a new one has been requested for the same phone number -> ONE_TIME_PASSWORD_SMS.VERIFICATION_EXPIRED
2. The authenticationId is invalid (i.e because the client sends invalid characters) -> ONE_TIME_PASSWORD_SMS.INVALID_OTP
3. The authenticationId is no longer valid because it has already been used -> ONE_TIME_PASSWORD_SMS.VERIFICATION_EXPIRED
4. Missing authenticationId in the request -> INVALID_ARGUMENT

**Comment by bigludo7** [comment]
Thanks @fernandopradocabrillo for the review.

I've added these scenarios but also I've reorganized the file to make it easier to read.
I have one point point: 403 did not seems to be relevant for validate-code operation. wdyt?

File: code/Test_Definitions/OTPvalidationAPI.feature [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+###########################
+#  401 errors for send_code
+###########################
+
+  @OTPvalidationAPI_401.1_send_code_no_authorization_header
+  Scenario: No Authorization header
+    Given the header "Authorization" is removed
+    And the resource "/one-time-password-sms/v0/send-code"
+    When the HTTP "POST" request is sent
+    Then the response property "$.status" is 403
```
Comment (trehman-gsma) [inline code/Test_Definitions/OTPvalidationAPI.feature:204]: Should this be a 401 instead of a 403?

File: code/Test_Definitions/OTPvalidationAPI.feature [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+    Then the response property "$.status" is 401
+    And the response property "$.code" is "UNAUTHENTICATED"
+    And the response property "$.message" contains a user friendly text
+
+
+###########################
+#  404 errors for send_code
+###########################
+
+  @OTPvalidationAPI_404.1_send_code_resource_not_found
```
Comment (trehman-gsma) [inline code/Test_Definitions/OTPvalidationAPI.feature:231]: Hi @bigludo7 - I think this might be challenging for some Operators to implement. Are there wider guidelines that state that sub-paths should be treated with CAMARA error handling?

I think some implementations will only expose only specific paths/routes in their API Gateway e.g. only `/one-time-password-sms/v0/send-code` is explicitly exposed as a CAMARA API, and anything else is considered an arbitrary URL. In this test case,  `/one-time-password-sms/send-code` might be considered an arbitrary URL and therefore CAMARA error handling is not applied.

File: code/Test_Definitions/OTPvalidationAPI.feature [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+    And the request body property "$.code" is set to a format valid value
+    And the resource "/one-time-password-sms/v0/validate-code"
+    When the HTTP "POST" request is sent
+    Then the response status code is 400
+    And the response property "$.status" is 400
+    And the response property "$.code" is "INVALID_ARGUMENT"
+    And the response property "$.message" contains a user friendly text
+    And the response header "x-correlator" has same value as the request header "x-correlator"
+
+  @OTPvalidationAPI_400.3_validate_code_missing_authenticationId
```
Comment (trehman-gsma) [inline code/Test_Definitions/OTPvalidationAPI.feature:395]: I think this is a duplicate title. Should it be named `@OTPvalidationAPI_400.3_validate_code_missing_code` instead?

File: code/Test_Definitions/OTPvalidationAPI.feature [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+    Given an authenticationId has been retrieved from a send-code request
+    And the request body property "$.code" is not valued
+    And the resource "/one-time-password-sms/v0/validate-code"
+    When the HTTP "POST" request is sent
+    Then the response status code is 400
+    And the response property "$.status" is 400
+    And the response property "$.code" is "INVALID_ARGUMENT"
+    And the response property "$.message" contains a user friendly text
+    And the response header "x-correlator" has same value as the request header "x-correlator"
+
```
Comment (trehman-gsma) [inline code/Test_Definitions/OTPvalidationAPI.feature:406]: What do you think about an additional case to test compliance with the maxLength of the code field? 
Example:
```
 @OTPvalidationAPI_400.4_validate_code_exceed_code_max_length
  Scenario: exceed the maxLength for code
    Given request body property "$.authenticationId" is set to the value from send-code request
    And the request body property "$.code" is set to "thisCodeExceedsTenCharacters"
    And the resource "/one-time-password-sms/v0/validate-code"
    When the HTTP "POST" request is sent
    Then the response status code is 400
    And the response property "$.status" is 400
    And the response property "$.code" is "INVALID_ARGUMENT"
    And the response property "$.message" contains a user friendly text
    And the response header "x-correlator" has same value as the request header "x-correlator"
```

File: code/Test_Definitions/OTPvalidationAPI.feature [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+    And the request body property "$.message" is longer than config_var:"max_lenght"
+    And the resource "/one-time-password-sms/v0/send-code"
+    When the HTTP "POST" request is sent
+    Then the response property "$.status" is 400
+    And the response property "$.code" is "INVALID_ARGUMENT"
+    And the response property "$.message" contains a user friendly text
+    And the response header "x-correlator" has same value as the request header "x-correlator" 
+
+###########################
+#  401 errors for send_code
```
Comment (trehman-gsma) [inline code/Test_Definitions/OTPvalidationAPI.feature:196]: Just an observation about the 401 test cases. There is no x-correlator validation in the 401 test cases whereas the other test cases in the document do check for x-correlator. 

On the same topic. Should there be happy path test cases that do not input an x-correlator? This would be to ensure that x-correlator is treated an optional header in the implementation.

Comment (bigludo7) [reply code/Test_Definitions/OTPvalidationAPI.feature:216]: I've added the x-correlator for 401

Adding testing without x-correlator could be added as this is not mandatory. I've added them. I did not test it in response as it is optional to provide one if not provided in the request.

File: code/Test_Definitions/OTPvalidationAPI.feature (hunk 1 above)
Comment (bigludo7) [reply code/Test_Definitions/OTPvalidationAPI.feature]: Yes ! fixed.

File: code/Test_Definitions/OTPvalidationAPI.feature (hunk 3 above)
Comment (bigludo7) [reply code/Test_Definitions/OTPvalidationAPI.feature]: Yes ! :)

File: code/Test_Definitions/OTPvalidationAPI.feature (hunk 4 above)
Comment (bigludo7) [reply code/Test_Definitions/OTPvalidationAPI.feature:430]: Yes !
Added

File: code/Test_Definitions/OTPvalidationAPI.feature (hunk 2 above)
Comment (bigludo7) [reply code/Test_Definitions/OTPvalidationAPI.feature:254]: Yes make sense. I've added a comment that this Test Case could be considered as optional. 
As for this API we target stable maturity level I tend to prefer to keep it as the testing guideline indicates: _Any public release must include a full test plan, covering both sunny and rainy day scenarios, to be considered stable. This may include unusual or not explicitly specified error scenarios, semantic validations, and corner cases._
We're in a corner case here probably

**Review by bigludo7 — COMMENTED** [review]
Thanks @trehman-gsma !!
This is very helpful.
I took into consideration all your comments.

File: code/Test_Definitions/OTPvalidationAPI.feature [hunk 6]
```diff
@@ -0,0 +1,528 @@
...
+    And the resource "/one-time-password-sms/v0/send-code"
+    When the HTTP "POST" request is sent
+    Then the response property "$.status" is 403
+    And the response property "$.code" is "ONE_TIME_PASSWORD_SMS.PHONE_NUMBER_NOT_ALLOWED"
+    And the response property "$.message" contains a user friendly text
+    And the response header "x-correlator" has same value as the request header "x-correlator" 
+
+@OTPvalidationAPI_05_send_code_phone_number_not_allowed_3
+  Scenario: Validation for failed scenario for a phone number that target a landline
+  Given the request body property "$.phoneNumber" is set to a phone number that target an landline
```
Comment (fernandopradocabrillo) [inline code/Test_Definitions/OTPvalidationAPI.feature:99]:
```suggestion
  Given the request body property "$.phoneNumber" is set to a phone number that target a landline

```

File: code/Test_Definitions/OTPvalidationAPI.feature [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+#  404 errors for validate_code
+###############################
+
+  @OTPvalidationAPI_404_validate_code_resource_not_found
+  Scenario: resource not found 
+      Given the request body property "$.authenticationId" is set to an unknown value
+      And the resource "/one-time-password-sms/validate-code"
+      When the HTTP "POST" request is sent
+      Then the response property "$.status" is 404
+      And the response property "$.code" is NOT_FOUND"
```
Comment (fernandopradocabrillo) [inline code/Test_Definitions/OTPvalidationAPI.feature:480]:
```suggestion
      And the response property "$.code" is "NOT_FOUND"

```

**Review by fernandopradocabrillo — APPROVED** [review]
LGTM

Thanks @bigludo7 @trehman-gsma !
